### ISSUE TITLE
Can't play specified track from playlist

### ISSUE THREAD

**Issue by kooverd** [olejon/spotcommander#114]
Hi!

I've encountered a small error. Nothing that I cannot live with, but I thought I'd let you know.

Spotcommander 13.2
Spotify /according to apt-get/ 1:1.0.49.125.g72ee7853-111

What I do:
Music is playing with no queue. I go to Playlist(*), select one, choose browse and select a single track than play it. "Now playing" is refreshing, but after that go back to the track that have been uninterruptedly playing.
(*) - that error occurred on every playlist that was longer than 308 tracks and never on ones that did not exceeded 108. I have yet to determine the border value.

By the way:
You've written "Playing a track in any list of tracks, ranging from Library to playlists, artists, albums, recently played etc will continue playing tracks in that list". What I observed, it more like Spotcommander plays that whole list from the beginning rather than continue playing it from previously selected place, even if one have chosen first track.
In other words, selecting 1 results in playing 1;1;2;3;4;5... selecting 3 results in 3;1;2;3;4;5... I'd expect 1;2;3;4;5... and 3;4;5;6;7;8... respectively. Though I don't know if it is a bug or feature.

Though you did really great work with the whole Spotcommander.

**Comment by olejon** [comment]
Thanks for the bug report. I'd expect bugs as there are many changes.

Important question: Do you have Spotify Premium? Looks like you do.

You're right about the playback order. It's a workaround since there is no queue API yet, so what it does when you play a track in a playlist for example is to add all the URIs to the new Player API, but before that puts the chosen track URI as the first URI in the array (list of tracks sent to Spotify server). If you do not have shuffle enabled, it will play as you describe yes. Would require some code for this behavior to change. One could remove the tracks above from the array, but then they won't play with shuffle enabled and won't play if repeat is enabled and the last track finishes. So maybe add them to the end of the array.

I always have shuffle enabled so never really encountered this when using it. That's why I encourage people to report an issue or contact me if they find anything, like you did :)

Thanks. I will look at this for the next version. Things are pretty ready now for compatibility with v 0.9 and v 1.0 of Spotify and Premium and Free users, so more minor stuff like this is not that much work I guess.

I also discovered a bug when adding SpotCommander to the home screen on iOS. It doesn't preserve the state as before and one has to open the menu to go to a page, and also manually refresh now playing it seems.

**Comment by olejon** [comment]
BTW you can still use the old behavior by disabling premium features in the file `/use/share/spotcommander/config.php`. But that works absolutely best with Spotify 0.9. Proper support for v 1.0 is only possible thanks to the new Player API, because they broke the OpenURI() MPRIS D-Bus in v 1.0, or changed its behavior, so the problem is that playing a track will repeat it over and over. On v 0.9 this is not a problem. Also the queue is probably more stable there with premium features disabled. The wiki has detailed instructions of how to install it.

**Comment by olejon** [comment]
Probably the easiest would be to use the offset key differently:

https://developer.spotify.com/web-api/start-a-users-playback/

**Comment by kooverd** [comment]
For the record: yes, I do have premium.

Turning shuffle on did not change the fact that I can't launch 300+ tracks playlist that way. My guess (without digging in the code) is that when playing track from long playlist, everything works as you've described, but then there's some kind of overflow (for example 300 tracks to be added and maximum for that array is 255 or 256), which causes an error and after all nothing is changed at all.

I'll stick to 1.0. I often use regular app to launch one of "Daily Mixes" from PC on server. When I had 0.9 on server and 1.0 on PC there was some trouble with them, as what should be endless playlist ended up being loop of about 30 tracks. I hope that that if I play them from 1.0 PC on 1.0 server that issue wile be gone.

**Comment by olejon** [comment]
Alright, thanks for the info. The limit that thing doesn't have to do with shuffle no. That's a shame, didn't try with a playlist with that many tracks I think. Need to confirm though. The API documentation doesn't say anything about it so maybe it's my code sending to much data PHP <> JavaScript.

Should be easy to test that part.

**Comment by kooverd** [comment]
Yes, you can play entire playlist, both in playlist menu and by that button on top of the track list while browsing it. That part work with no problem even with 1000+ lists.

**Comment by olejon** [comment]
Ok, thanks.

**Comment by olejon** [comment]
Ok I have at least confirmed the limit problem with a playlist I have with more than 500 tracks. Will check on my computer later if the limit is in my code or the API. If it's in the API there is nothing I can do about it as far as I can think. The number of tracks you a PUT (HTTP request) to the API is a one time thing, you can't PUT more tracks like if it's a queue. If it's my code it should be possible to solve. The old solution which works with v 0.9 and no Premium supports as many tracks you want in the queue as it simply reads from the queue SQLite database file and plays the next track and removes it from the database. This unfortunately doesn't work with v 1.0 as OpenURI() in MPRIS just repeats a track URI passed to it over and over, and then the dbus-monitor process started by the daemon to get a signal when a track changes (to immediately play the next track in the queue instead of the track the Spotify client wants to play) doesn't get a signal of a new track playing from D-Bus.

**Comment by olejon** [comment]
Ok, I 've debugged the problem, and it is unfortunately the API. Read my issue report here for details:

https://github.com/spotify/web-api/issues/582

Any workaround with be nice. The data is sent minimized without spaces.

**Comment by olejon** [comment]
Ok, so I have at least succeeded in fixing the playback order problem. If shuffle is off, it now works as expected in my code (not released yet). So if you play track 8, it will continue to play 9, 10 and so on after that when using the play next track feature or when the track finishes.

This is possible passing the track's URI in the `offset` > `uri` JSON data in the Player API.

**Comment by olejon** [comment]
Order fixed in v 13.3. The limit I can not control, but at least a dialog is shown with two options.

**Comment by kooverd** [comment]
What if, in long playlist, your program will play the whole playlist from the start, and then sent appropriate number of "next track" commands? That will of course work only with disabled shuffle. Main question is if it is possible to skip 100s of tracks during couple of seconds. If not - that sugesstion is pointless.

**Comment by olejon** [comment]
No, a PUT to the API is basically telling Spotify to play put tracks next, in order if shuffle is disabled of random if not. You can think at it as a queue. When you PUT something to the API, if it's a context_uri (artists, albums and playlists) or uris (read the API doc) it replaces what have been put before. It's a shame one can't send `context_uri` and a `uris` to the API at the same time (it does only accept one of them). Because if so the standard playback would be easy. Just PUT the playlist URI as `context_uri` and the track in `uris`, then after playing the track, the playlist would play. That would also solve the wish many have for a queue API.

But I have modified the queue behavior. Because everything you queue in SpotCommander is added to a SQLite database on the computer, and I've added `LIMIT 226` to the SQL query to the database. So even though there are 1000 tracks only the 226 first tracks will be put to the API and every time a track plays it removes it from the database. So I think the workaround would be to queue all tracks when playing a single track in a playlist. That should work. With a dialog asking if one want to queue the tracks randomly or in order.

That feature was in SpotCommander 13.1 and older, from the overflow menu (but those versions don't support all the new stuff). So I can look into that for the next version. It's not a big change. But I had to get v 13.3 out now, especially because of iOS not working properly etc.

**Comment by olejon** [comment]
Just edited the above for clarifying. See you also did the same :)

**Comment by olejon** [comment]
Discovered that Spotify has increased this limit to `378`. The next version will contain this limit in the `config.php` file so if they increase it further you can easily change it without changing various files. As of now, you can go to the directory `spotcommander` (cd `/usr/share/spotcommander`) if you installed the deb package and run `grep -RI 226 *` and change the files with the number to `387` (except the cache files that may show up).

**Comment by olejon** [comment]
Increased to `378` in v 13.4. This limit can now easily be changed in `config.php` so if they increase the limit more one can easily edit it before I release a new version.

Files are in `/usr/share/spotcommander` if you installed the deb package.
